src/tls_relation.py

```diff
@@ -5,200 +5,169 @@
 """Gateway API TLS relation business logic."""
 import secrets
 import string
-from typing import Dict, Union
+import typing
 
 from charms.tls_certificates_interface.v3.tls_certificates import (
-    CertificateAvailableEvent,
     CertificateExpiringEvent,
     CertificateInvalidatedEvent,
+    ProviderCertificate,
     TLSCertificatesRequiresV3,
     generate_csr,
     generate_private_key,
 )
 from cryptography import x509
 from cryptography.x509.oid import NameOID
-from ops.jujuversion import JujuVersion
 from ops.model import Model, Relation, SecretNotFoundError
 
 TLS_CERT = "certificates"
 
 
-class TLSRelationService:
-    """TLS Relation service class."""
+class InvalidCertificateError(Exception):
+    """Exception raised when certificates is invalid."""
 
-    def __init__(self, model: Model) -> None:
-        """Init method for the class.
 
-        Args:
-            model: The charm model used to get the relations and secrets.
-        """
-        self.charm_model = model
-        self.charm_app = model.app
+class KeyPair(typing.NamedTuple):
+    """Stores a private key and encryption password.
 
-    def generate_password(self) -> str:
-        """Generate a random 12 character password.
+    Attrs:
```

Comment:
I think `Attrs` doesn't work with docstrings-complete. `Attributes` should work though :D

Reply:
src-docs doesn't render it properly if its named `Attrs`

Reply:
I renamed all `Attrs` to `Attributes`